types/wlr_input_method_v2.c

```diff
@@ -101,10 +119,177 @@ static void im_get_input_popup_surface(struct wl_client *client,
+		MAP_SHARED, keymap_fd, 0);
+	if (ptr == MAP_FAILED) {
+		wlr_log(WLR_ERROR, "failed to mmap() %zu bytes",
+			keyboard->keymap_size);
+		close(keymap_fd);
+		return false;
+	}
+
+	strcpy(ptr, keyboard->keymap_string);
+	munmap(ptr, keyboard->keymap_size);
+
+	zwp_input_method_keyboard_grab_v2_send_keymap(im_keyboard_grab->resource,
+		WL_KEYBOARD_KEYMAP_FORMAT_XKB_V1, keymap_fd,
+		keyboard->keymap_size);
+
+	close(keymap_fd);
+	return true;
+}
+
+static void im_keyboard_grab_send_repeat_info(
+		struct wlr_input_method_keyboard_grab_v2 *im_keyboard_grab,
+		struct wlr_keyboard *keyboard) {
+	zwp_input_method_keyboard_grab_v2_send_repeat_info(
+		im_keyboard_grab->resource, keyboard->repeat_info.rate,
+		keyboard->repeat_info.delay);
+}
+
+static void handle_keyboard_keymap(struct wl_listener *listener, void *data) {
+	struct wlr_input_method_keyboard_grab_v2 *im_keyboard_grab =
+		wl_container_of(listener, im_keyboard_grab, keymap_listener);
+	im_keyboard_grab_send_keymap(im_keyboard_grab, data);
+}
+
+static void handle_keyboard_repeat_info(struct wl_listener *listener,
+		void *data) {
+	struct wlr_input_method_keyboard_grab_v2 *im_keyboard_grab =
+		wl_container_of(listener, im_keyboard_grab, keymap_listener);
+	im_keyboard_grab_send_repeat_info(im_keyboard_grab, data);
+}
+
+void wlr_input_method_keyboard_grab_v2_set_keyboard(
+		struct wlr_input_method_keyboard_grab_v2 *im_keyboard_grab,
+		struct wlr_keyboard *keyboard) {
+	if (keyboard == im_keyboard_grab->keyboard) {
+		return;
+	}
+
+	if (im_keyboard_grab->keyboard) {
+		wl_list_remove(&im_keyboard_grab->keymap_listener.link);
+		wl_list_remove(&im_keyboard_grab->repeat_info_listener.link);
+	}
+
+	if (im_keyboard_grab->keyboard == NULL ||
+			strcmp(im_keyboard_grab->keyboard->keymap_string,
+			keyboard->keymap_string) != 0) {
+		// send keymap only if it is changed, or if input method is not
+		// aware that it did not change and blindly send it back with
+		// virtual keyboard, it may cause an infinite recursion.
+		if (!im_keyboard_grab_send_keymap(im_keyboard_grab, keyboard)) {
+			wlr_log(WLR_ERROR, "Failed to send keymap for input-method keyboard grab");
+			return;
+		}
+	}
+	im_keyboard_grab_send_repeat_info(im_keyboard_grab, keyboard);
+
+	im_keyboard_grab->keyboard = keyboard;
+	im_keyboard_grab->keymap_listener.notify = handle_keyboard_keymap;
+	wl_signal_add(&keyboard->events.keymap,
+		&im_keyboard_grab->keymap_listener);
+	im_keyboard_grab->repeat_info_listener.notify =
+		handle_keyboard_repeat_info;
+	wl_signal_add(&keyboard->events.repeat_info,
+		&im_keyboard_grab->repeat_info_listener);
+};
 
 static void im_grab_keyboard(struct wl_client *client,
 		struct wl_resource *resource, uint32_t keyboard) {
-	wlr_log(WLR_INFO, "Stub: zwp_input_method_v2::grab_keyboard");
+	struct wlr_input_method_v2 *input_method =
+		input_method_from_resource(resource);
+	if (!input_method) {
+		return;
+	}
+	if (input_method->im_keyboard_grab) {
+		// Already grabbed
```

Comment:
Should we send a client error here? If it's missing, we should add it to the protocol.